File: spark-scala/src/main/scala/com/tiger/spark/_1_word_count/WordCount.scala
package com.tiger.spark._1_word_count

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

object WordCount {

  def main(args: Array[String]): Unit = {

    // 1. 创建spark config
    // 这里的local表示使用本地线程模拟分布式, 默认为1个线程, 可以使用local[2]表示使用两个线程来模拟, local[*]表示使用当前cpu核数的线程数来模拟
    val sparkConf: SparkConf = new SparkConf().setMaster("local").setAppName("WordCount")

    // 2. 创建spark上下文
    val context: SparkContext = new SparkContext(sparkConf)

    // 3. 通过textFile按行读取文件
    val fildRDD: RDD[String] = context.textFile("input/")

    // 4. 将行通过空格进行分词
    val words: RDD[String] = fildRDD.flatMap(_.split(" "))

    // 5. 将单词转换为元组
    val tuple: RDD[(String, Int)] = words.map((_, 1))

    // 6. 将元组按照key进行分组, 这里进行了隐式转换, 将RDD类型转换为PairRDDFunctions
    val reduce: RDD[(String, Int)] = tuple.reduceByKey(_ + _)

    // 7. 将聚合结果采集到内存
    val result: Array[(String, Int)] = reduce.collect()

    // 8. 打印结果
    result.foreach(println)

    // 9. 关闭连接
    context.stop()





    
  }
}

File: spark-scala/src/main/java/_1_word_count/WordCount.java
package _1_word_count;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function2;
import scala.Tuple2;

import java.util.Arrays;

/**
 * 该测试用例可以远程提交spark程序
 */
public class WordCount {

    /**
     * 没有隐式转换是真的难用, spark的java api
     * 这套api就是根本就不是原生为java设计的, 只是在scala api上面对java适配
     */
    public static void main(String[] args) {
        SparkConf conf = new SparkConf()
            .setAppName("java_wordcount")
            .setMaster("spark://hadoop102:7077")
            .setJars(new String[]{"C:\\Users\\Administrator\\Desktop\\example4all\\spark-scala\\target\\spark-scala-1.0-SNAPSHOT.jar"});
        // 设置hdfs用户
        System.setProperty("HADOOP_USER_NAME", "tiger");

        JavaSparkContext context = new JavaSparkContext(conf);

        JavaRDD<String> textFile = context.textFile("hdfs://hadoop102:8020/wordcount/data.txt");

        JavaRDD<String> flatMap = textFile.flatMap(str -> Arrays.asList(str.split(" ")).iterator());

        // 这里因为没有scala的隐式转换, 所以需要调用toPair将一个单value类型转换为一个key-value类型, 才会有后面的reduceByKey方法
        JavaPairRDD<String, Integer> tupleRDD = flatMap.mapToPair(str -> new Tuple2<>(str, 1));

        // 这里必须是JavaPairRDD才有reduceByKey这个方法, JavaRDD中没有这个方法
        JavaPairRDD<String, Integer> reduce = tupleRDD.reduceByKey(Integer::sum);

        // 因为JavaPairRDD只有sortByKey这个方法, 没有sortBy这个方法(sortBy这个方法在JavaRDD中),
        // 所以必须调换key和value的位置, 才能根据count来排序
        JavaPairRDD<Integer, String> map = reduce.mapToPair(Tuple2::swap);

        JavaPairRDD<Integer, String> sort = map.sortByKey(false);
        sort.saveAsTextFile("hdfs://hadoop102:8020/wordcount/output11111");

        context.stop();
    }
}
